vm/src/stdlib/ctypes/basics.rs

```diff
@@ -0,0 +1,316 @@
+        buffer_copy(cls, obj, offset, vm, true)
+    }
+
+    #[pyclassmethod]
+    fn in_dll(
+        cls: PyTypeRef,
+        dll: PyObjectRef,
+        name: PyStrRef,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyCData> {
+        if let Ok(h) = vm.get_attribute(cls.as_object().to_owned(), "_handle") {
+            // This is something to be "CPython" like
+            let raw_ptr = if let Ok(h_int) = h.downcast_exact::<PyInt>(vm) {
+                dlsym(h_int, name, vm)
+            } else {
+                Err(vm.new_type_error(format!(
+                    "_handle must be an int not {}",
+                    dll.clone().class().name
+                )))
+            }?;
+
+            let sym_ptr = usize::try_from_object(vm, raw_ptr)?;
+
+            let buffer = at_address(&cls, sym_ptr, vm)?;
+            Ok(PyCData::new(None, Some(buffer)))
+        } else {
+            Err(vm.new_attribute_error("atribute '_handle' not found".to_string()))
+        }
+    }
+}
+
+#[pyimpl]
+pub trait PyCDataSequenceMethods: PyValue {
+    // CDataType_as_sequence methods are default for all *Type_Type
+    // Basically the sq_repeat slot is CDataType_repeat
+    // which transforms into a Array
+
+    // #[pymethod(name = "__mul__")]
+    // fn mul(&self, counter: isize, vm: &VirtualMachine) -> PyObjectRef {
+    // }
+
+    // #[pymethod(name = "__rmul__")]
+    // fn rmul(&self, counter: isize, vm: &VirtualMachine) -> PyObjectRef {
+    //     self.mul(counter, vm)
+    // }
+}
+
+impl<'a> BorrowValue<'a> for PyCData {
+    type Borrowed = PyRwLockReadGuard<'a, Vec<u8>>;
+
+    fn borrow_value(&'a self) -> Self::Borrowed {
+        self._buffer.read()
+    }
+}
+
+impl BufferProtocol for PyCData {
+    fn get_buffer(zelf: &PyRef<Self>, _vm: &VirtualMachine) -> PyResult<Box<dyn Buffer>> {
+        Ok(Box::new(PyCDataBuffer {
+            data: zelf.clone(),
+            options: BufferOptions {
+                readonly: false,
+                len: zelf._buffer.read().len(),
+                ..Default::default()
+            },
+        }))
+    }
+}
+
+#[derive(Debug)]
+struct PyCDataBuffer {
+    data: PyCDataRef,
+    options: BufferOptions,
+}
+
+// This trait will be used by all types
+impl Buffer for PyCDataBuffer {
+    fn obj_bytes(&self) -> BorrowedValue<[u8]> {
+        PyRwLockReadGuard::map(self.data.borrow_value(), |x| x.as_slice()).into()
+    }
+
+    fn obj_bytes_mut(&self) -> BorrowedValueMut<[u8]> {
+        PyRwLockWriteGuard::map(self.data.borrow_value_mut(), |x| x.as_mut_slice()).into()
+    }
+
+    fn release(&self) {}
+
+    fn get_options(&self) -> &BufferOptions {
+        &self.options
+    }
+}
+
+// This Trait is the equivalent of PyCData_Type on tp_base for
+// Struct_Type, Union_Type, PyCPointer_Type
+// PyCArray_Type, PyCSimple_Type, PyCFuncPtr_Type
+#[pyclass(module = "_ctypes", name = "_CData")]
+pub struct PyCData {
+    _objects: AtomicCell<Vec<PyObjectRef>>,
+    _buffer: PyRwLock<Vec<u8>>,
```

Comment:
I think the buffer is supposed to be a raw pointer and a length, so that you can mutate the original data it if you want to.

Reply:
Something like
```rust
struct RawBuffer{
    inner: *mut u8,
    size: usize
}
```
?